File: 2.3 Nonlinear DS/Set/p978.py
# Competitive Programming 3
# Problem 978

# note - this code is for testing only, does not use appropriate datastructure
# for fast insertion/deletion

T = int(input())

for t in range(T):
    B, SG, SB = [int(x) for x in input().split()]
    
    green, blue = [0 for _ in range(SG)], [0 for _ in range(SB)]
    for i in range(SG): green[i] = int(input())
    for i in range(SB): blue[i] = int(input())
    
    green.sort(reverse=True)
    blue.sort(reverse=True)
    
    
    while SB > 0 and SG > 0:
        numFights = 0
        for i in range(B):
            if i + 1 > SB or i + 1 > SG: break
            numFights += 1
        
            fight = min(green[i], blue[i])
            green[i], blue[i] = green[i] - fight, blue[i] - fight
        
        for i in range(numFights-1,-1,-1):
            if green[i] == 0: 
                green.pop(i)
                SG -= 1
            if blue[i] == 0: 
                blue.pop(i)
                SB -= 1
    
        green.sort(reverse=True)
        blue.sort(reverse=True)
            
    if len(blue) == 0 and len(green) == 0:
        print("green and blue died")
    elif len(blue) == 0:
        print("green wins")
        print(*green, sep='\n')
    else:
        print("blue wins")
        print(*blue, sep='\n')
        
    if t < T-1: print('')
        
    

File: 2.3 Nonlinear DS/Set/p978.cpp
// Competitive Programming 3
// Problem 978

#include <algorithm>
#include <cstring>
#include <iostream>
#include <map>
#include <set>
#include <string>
using namespace std;

int main(){
    int T;
    cin >> T;

    while (T--){
        int B, SG, SB, n;
        cin >> B >> SG >> SB;

        multiset <int,  greater<int> > blue;
        multiset <int,  greater<int> > green;
        set<int>::iterator it_green;
        set<int>::iterator it_blue;
        int greenFighters[B], blueFighters[B];

        for (int i = 0; i < SG; i++){
            cin >> n;
            green.insert(n);
        }
        for (int i = 0; i < SB; i++){
            cin >> n;
            blue.insert(n);
        }

        while (SB > 0 && SG > 0){
            int numFights = min(B, min((int)green.size(), (int)blue.size()));
            it_green = green.begin();
            it_blue = blue.begin();

            for (int i = 0; i < numFights; i++){
                greenFighters[i] = *it_green;
                green.erase(it_green++);
                blueFighters[i] = *it_blue;
                blue.erase(it_blue++);

                int hit = min(greenFighters[i], blueFighters[i]);
                greenFighters[i] -= hit;
                blueFighters[i] -= hit;
            }

            for (int i = 0; i < numFights; i++){
                if (greenFighters[i] > 0)
                    green.insert(greenFighters[i]);
                else
                    SG -= 1;

                if (blueFighters[i] > 0)
                    blue.insert(blueFighters[i]);
                else
                    SB -= 1;
            }
        }

        if (SG > 0){
            cout << "green wins" << endl;
            for (it_green = green.begin(); it_green!= green.end(); it_green++)
                cout << *it_green << endl;
        }
        else if (SB > 0){
            cout << "blue wins" << endl;
            for (it_blue = blue.begin(); it_blue!= blue.end(); it_blue++)
                cout << *it_blue << endl;
        }
        else
            cout << "green and blue died" << endl;

        if (T > 0) cout << endl;
    }
    return 0;
}
